src/rust-crypto/verification.ts

```diff
@@ -451,15 +456,16 @@ abstract class BaseRustVerifer<InnerType extends RustSdkCryptoJs.Qr | RustSdkCry
 > {
     /** A promise which completes when the verification completes (or rejects when it is cancelled/fails) */
     protected readonly completionPromise: Promise<void>;
+    private onChangeCallback!: () => Promise<void>;
 
     public constructor(
-        protected readonly inner: InnerType,
+        protected inner: InnerType,
         protected readonly outgoingRequestProcessor: OutgoingRequestProcessor,
     ) {
         super();
 
         this.completionPromise = new Promise<void>((resolve, reject) => {
-            const onChange = async (): Promise<void> => {
+            this.onChangeCallback = async (): Promise<void> => {
```

Comment:
I'm looking at this again and feeling like we're jumping through hoops here. Suggestion:

 * Replace `completionPromise` with a deferred (`defer()` in `js-sdk/src/utils.ts`). Critically, that means that we can hoist this code out of the closure.
 * Turn `onChangeCallback` into a regular async method. Combine it with `onChange`; the subclasses can override it and call `super.onChange()` in the overriding impl.

Reply:
Ah, I didn't know we had a `defer`. :+1:

Reply:
OK, I tried doing so, but tests are failing, so I must be doing something wrong.  It's giving errors like:

```
  ● verification (rust-sdk) › cancellation › can cancel in the ShowQrCodeCallbacks

    RuntimeError: unreachable

      at std::panicking::rust_panic_with_hook::h366c91ab824150d6 (wasm:/wasm/014eee0e:1:2591137)
      at core::panicking::panic_fmt::h9728c4a03fdab7ce (wasm:/wasm/014eee0e:1:3444709)
      at matrix_sdk_crypto_wasm::machine::promise_result_to_future::{{closure}}::hb5a608792e59a66e (wasm:/wasm/014eee0e:1:1744633)
      at matrix_sdk_crypto_wasm::verification::send_change_info_to_callback::{{closure}}::hae64d2c5c9c49bef (wasm:/wasm/014eee0e:1:1394630)
      at matrix_sdk_crypto_wasm::verification::Qr::register_changes_callback::{{closure}}::ha47e9b5b026dda4b (wasm:/wasm/014eee0e:1:1750252)
      at wasm_bindgen_futures::queue::Queue::new::{{closure}}::h7ef0740be65eb2e0 (wasm:/wasm/014eee0e:1:2130242)
      at wasm_bindgen::convert::closures::invoke1_mut::h638884e1ee5d047c (wasm:/wasm/014eee0e:1:3559764)
      at wasm_bindgen__convert__closures__invoke1_mut__h638884e1ee5d047c (../matrix-rust-sdk-crypto-wasm/pkg/matrix_sdk_crypto_wasm.js:275:10)
      at f (../matrix-rust-sdk-crypto-wasm/pkg/matrix_sdk_crypto_wasm.js:217:20)

  ● verification (rust-sdk) › cancellation › can cancel in the ShowQrCodeCallbacks

    thrown: "Exceeded timeout of 5000 ms for a test.
    Add a timeout value to this test to increase the timeout, if this is a long-running test. See https://jestjs.io/docs/api#testname-fn-timeout."

      787 |         });
      788 |
    > 789 |         it("can cancel in the ShowQrCodeCallbacks", async () => {
          |         ^
      790 |             // have alice initiate a verification. She should send a m.key.verification.request
      791 |             const [, request] = await Promise.all([
      792 |                 expectSendToDeviceMessage("m.key.verification.request"),

      at it (spec/integ/crypto/verification.spec.ts:789:9)
      at describe (spec/integ/crypto/verification.spec.ts:716:5)
          at Array.forEach (<anonymous>)
      at Object.<anonymous> (spec/integ/crypto/verification.spec.ts:119:47)
```

The patch I currently have is:

```diff
diff --git a/src/rust-crypto/verification.ts b/src/rust-crypto/verification.ts
index a163c3a0d..18f9d0025 100644
--- a/src/rust-crypto/verification.ts
+++ b/src/rust-crypto/verification.ts
@@ -34,6 +34,7 @@ import { OutgoingRequest, OutgoingRequestProcessor } from "./OutgoingRequestProc
 import { TypedReEmitter } from "../ReEmitter";
 import { MatrixEvent } from "../models/event";
 import { EventType, MsgType } from "../@types/event";
+import { defer, IDeferred } from "../utils";
 
 /**
  * An incoming, or outgoing, request to verify a user or a device via cross-signing.
@@ -455,8 +456,7 @@ abstract class BaseRustVerifer<InnerType extends RustSdkCryptoJs.Qr | RustSdkCry
     VerifierEventHandlerMap & VerificationRequestEventHandlerMap
 > {
     /** A promise which completes when the verification completes (or rejects when it is cancelled/fails) */
-    protected readonly completionPromise: Promise<void>;
-    protected onChangeCallback!: () => Promise<void>;
+    protected readonly completionDeferred: IDeferred<void>;
 
     public constructor(
         protected inner: InnerType,
@@ -464,29 +464,29 @@ abstract class BaseRustVerifer<InnerType extends RustSdkCryptoJs.Qr | RustSdkCry
     ) {
         super();
 
-        this.completionPromise = new Promise<void>((resolve, reject) => {
-            this.onChangeCallback = async (): Promise<void> => {
-                this.onChange();
-
-                if (this.inner.isDone()) {
-                    resolve(undefined);
-                } else if (this.inner.isCancelled()) {
-                    const cancelInfo = this.inner.cancelInfo()!;
-                    reject(
-                        new Error(
-                            `Verification cancelled by ${
-                                cancelInfo.cancelledbyUs() ? "us" : "them"
-                            } with code ${cancelInfo.cancelCode()}: ${cancelInfo.reason()}`,
-                        ),
-                    );
-                }
-
-                this.emit(VerificationRequestEvent.Change);
-            };
-            inner.registerChangesCallback(this.onChangeCallback);
-        });
+        this.completionDeferred = defer();
+        inner.registerChangesCallback(() => { this.onChangeCallback() });
         // stop the runtime complaining if nobody catches a failure
-        this.completionPromise.catch(() => null);
+        this.completionDeferred.promise.catch(() => null);
+    }
+
+    protected async onChangeCallback(): Promise<void> {
+        this.onChange();
+
+        if (this.inner.isDone()) {
+            this.completionDeferred.resolve(undefined);
+        } else if (this.inner.isCancelled()) {
+            const cancelInfo = this.inner.cancelInfo()!;
+            this.completionDeferred.reject(
+                new Error(
+                    `Verification cancelled by ${
+                        cancelInfo.cancelledbyUs() ? "us" : "them"
+                     } with code ${cancelInfo.cancelCode()}: ${cancelInfo.reason()}`,
+                ),
+            );
+        }
+
+        this.emit(VerificationRequestEvent.Change);
     }
 
     /**
@@ -579,7 +579,7 @@ export class RustQrCodeVerifier extends BaseRustVerifer<RustSdkCryptoJs.Qr> impl
             this.emit(VerifierEvent.ShowReciprocateQr, this.callbacks);
         }
         // Nothing to do here but wait.
-        await this.completionPromise;
+        await this.completionDeferred.promise;
     }
 
     /**
@@ -657,7 +657,7 @@ export class RustSASVerifier extends BaseRustVerifer<RustSdkCryptoJs.Sas> implem
      */
     public async verify(): Promise<void> {
         await this.sendAccept();
-        await this.completionPromise;
+        await this.completionDeferred.promise;
     }
 
     /**
@@ -733,7 +733,7 @@ export class RustSASVerifier extends BaseRustVerifer<RustSdkCryptoJs.Sas> implem
     public replaceInner(inner: RustSdkCryptoJs.Sas): void {
         if (this.inner != inner) {
             this.inner = inner;
-            inner.registerChangesCallback(this.onChangeCallback);
+            inner.registerChangesCallback(() => { this.onChangeCallback() });
             // replaceInner will only get called if we started the verification at the same time as the other side, and we lost
             // the tie breaker.  So we need to re-accept their verification.
             this.sendAccept();
```

Reply:
Done, except for combining with `onChange`, because I'm not sure we want to rely on remembering to call `super.onChange` when we add a new verification method.

Reply:
> because I'm not sure we want to rely on remembering to call `super.onChange` when we add a new verification method.

I feel like:
 * if I were implementing a new subclass of `BaseRustVerifer`, in which I chose to override a method in that superclass... then I would remember to check if the superclass method needed calling
 * that's a pretty reasonable thing to expect of any future maintainer
 * Right now, there's nothing stopping me, as someone implementing a new verification method, from overriding `onChangeCallback` instead of `onChange` and messing things up just as badly, so any potential benefit is a bit of a placebo anyway
 * Having separate `onChangeCallback` and `onChange` methods makes the flow harder to understand

But... I'll not insist if you'd prefer to keep it this way.